SDK/MobileCenterRum/Microsoft.Azure.Mobile.Rum.Windows.Shared/RealUserMeasurements.cs

```diff
@@ -0,0 +1,381 @@
+                    MobileCenterLog.Error(LogTag, "Rum key is invalid.");
+                    return;
+                }
+                _rumKey = rumKey;
+            }
+        }
+
+        public override string ServiceName => "RealUserMeasurements";
+
+        protected override string ChannelName => "rum";
+
+        public override void OnChannelGroupReady(IChannelGroup channelGroup, string appSecret)
+        {
+            lock (_serviceLock)
+            {
+                base.OnChannelGroupReady(channelGroup, appSecret);
+                ApplyEnabledState(InstanceEnabled);
+            }
+        }
+
+
+        private void ApplyEnabledState(bool enabled)
+        {
+            lock (_serviceLock)
+            {
+                if (enabled)
+                {
+                    // Snapshot RUM key for this run, updates will work only if disabling/enabling again.
+                    var rumKey = _rumKey;
+                    if (rumKey == null)
+                    {
+                        MobileCenterLog.Error(LogTag, "Rum key must be configured before start.");
+                        return;
+                    }
+
+                    // Create cancellation token and snapshot token for the task to avoid race conditions.
+                    _cancellationTokenSource = new CancellationTokenSource();
+                    var cancellationToken = _cancellationTokenSource.Token;
+                    Task.Run(async () =>
+                    {
+                        try
+                        {
+                            await RunTestsAsync(rumKey, cancellationToken);
+                        }
+                        catch (OperationCanceledException)
+                        {
+                            MobileCenterLog.Debug(LogTag, "Measurements were canceled.");
+                        }
+                        catch (Exception e)
+                        {
+                            MobileCenterLog.Error(LogTag, "Could not run tests.", e);
+                        }
+                        finally
+                        {
+                            lock (_serviceLock)
+                            {
+                                if (cancellationToken == _cancellationTokenSource?.Token)
+                                {
+                                    DisposeCancellationSource();
+                                }
+                            }
+                        }
+                    }, cancellationToken);
+                }
+                else
+                {
+                    _cancellationTokenSource?.Cancel();
+                    DisposeCancellationSource();
+                }
+            }
+        }
+
+        private void DisposeCancellationSource()
+        {
+            _cancellationTokenSource?.Dispose();
+            _cancellationTokenSource = null;
+        }
+
+        private async Task RunTestsAsync(string rumKey, CancellationToken cancellationToken)
```

Comment:
General feedback - this method is too long, we need to refactor it in the future.

Reply:
Done in #ab6c7b1

Reply:
Looks fine to me.